keps/sig-node/5055-device-taints-and-tolerations/README.md

```diff
@@ -0,0 +1,978 @@
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+With Dynamic Resource Allocation (DRA), DRA drivers publish information about
+the devices that they manage in ResourceSlices. This information is used by the
+scheduler when selecting devices for user requests in ResourceClaims.
+
+With this KEP, DRA drivers can mark devices as tainted such that they won't be
```

Comment:
Should a device be provisioned before applying the taint? Will this support tainting partitionable devices?

Reply:
It's possible to set taints for devices that aren't in any ResourceSlice yet ("not provisioned").

Partitionable devices are still devices with names and attributes, so those also can be tainted. Extending the logic so that tainting a source device prevents allocating a not-tainted leaf device (is that the right term?) would be something that has to be implemented separately.

cc @mortent

Reply:
Do you know if the admin must keep device taints in sync with node taints? Is it an admin taint policy issue where the node taints are restrictive, but devices don't have any taints?

Reply:
Those two kind of taints are completely independent.

An admin can use node taints to affect all pods on a node, or device taints to only affect pods using devices. If both are set, both must be tolerated in the pod resp. claim for a pod to get scheduled or keep running.

Reply:
May be we can add the above discussion in KEP

Reply:
> Those two kind of taints are completely independent.

Yes, that's the point of KEP-5055. This KEP enables cluster admin taint per device (exclude (probably) failure device from scheduling/executing). This feature can improve device utlization of the cluster. Currently, 1 of 8 GPUs has failure and device plugin does not detected, cluster admin taint a whole node even though 7 GPUs are healthy. But, with this KEP, cluster admin can taint the exact failure GPU. 

> May be we can add the above discussion in KEP

I think it's reasonable. Currently, _Taint_ might be used as _Node Taint_ because no othor taints exist in Kubernetes.

Reply:
I added one paragraph under "Design Details".